stream_alert/shared/alert.py

```diff
@@ -0,0 +1,344 @@
+"""
+from datetime import datetime
+import json
+import uuid
+
+
+class Alert(object):
+    """Encapsulates a single alert and handles serializing to Dynamo and merging."""
+
+    class AlertEncoder(json.JSONEncoder):
+        """Custom JSON encoder which handles sets."""
+        def default(self, obj):  # pylint: disable=arguments-differ,method-hidden
+            if isinstance(obj, set):
+                return list(obj)
+            return json.JSONEncoder.default(self, obj)
+
+    _EXPECTED_INIT_KWARGS = {
+        'alert_id', 'attempts', 'cluster', 'context', 'created', 'dispatched', 'log_source',
+        'log_type', 'merge_by_keys', 'merge_window_mins', 'retry_outputs', 'rule_description',
+        'source_entity', 'source_service', 'staged'
+    }
+
+    def __init__(self, rule_name, record, outputs, **kwargs):
+        """Create a new Alert with a random ID and timestamped now.
+
+        Args:
+            rule_name (str): Name of the rule which triggered the alert
+            record (dict): Parsed log which triggered the alert
+            outputs (set): Set of string outputs where the alert should be delivered to
+
+        Kwargs (optional):
+            alert_id (str): Existing AlertID, if known. If not specified, a random one is created.
+            attempts (int): Number of attempts to deliver this alert so far. Defaults to 0.
+            cluster (str): Cluster from which the alert was generated.
+            context (dict): Context dictionary associated with the triggering rule.
+            created (str): Human-readable creation timestamp. If not specified, uses current time.
+            dispatched (int): UNIX timestamp when the alert was last dispatched for processing.
+            log_source (str): Log type which triggered the alert, e.g. "binaryalert"
+            log_type (str): The type of the triggering log. Usually "json"
+            merge_by_keys (list): Alerts are merged if the values associated with all of these
+                top-level keys are equal. TODO: Support nested merge keys
+            merge_window_mins (int): Merged alerts are sent at this interval.
+            retry_outputs (set): Subset of outputs which failed to send and should be retried.
+            rule_description (str): Description associated with the triggering rule.
+            source_entity (str): Name of location from which the record originated. E.g. "mychannel"
+            source_service (str): Input type from which the record originated. E.g. "slack"
+            staged (bool): Whether this rule is currently in the staging process. Defaults to False
+        """
+        if not set(kwargs).issubset(self._EXPECTED_INIT_KWARGS):
+            raise TypeError(
+                'Invalid Alert kwargs: {} is not a subset of {}'.format(
+                    ','.join(sorted(kwargs)), ','.join(sorted(self._EXPECTED_INIT_KWARGS))
+                )
+            )
+
+        self.alert_id = kwargs.get('alert_id') or str(uuid.uuid4())
```

Comment:
I address this in a comment [a few lines down](https://github.com/airbnb/streamalert/pull/666/files#diff-eb2d14d1777a27b0a3eb48839eaf8e83R77), but the basic problem is that you could have Falsey values of the wrong type.

In particular, Dynamo will reject empty strings and empty sets. So if someone were to call `Alert(alert_id='')`, a simple `kwargs.get(alert_id)` will accept the empty string value, which we don't want. The idea of using `or` is to force *any* Falsey value (0, None, '', [], {}) to the Falsey value of the appropriate type.

Happy to discuss whether this is necessary in every case (`alert_id` in particular should never be an empty string, but `cluster` or `log_source` could be), but it seemed a bit safer to enforce the correct types from the beginning of Alert creation

Reply:
aha! this is actually the reasoning I had behind [this code](https://github.com/airbnb/streamalert/blob/32f8825f19add9d8c49f2df392d2ac76e9ea253f/stream_alert_cli/terraform/athena.py#L45-L47) awhile back, but removed that because I couldn't recall my reasoning :LOL: the gist there was that I wanted to support having a value in the config (so the user knew the config setting existed) but if it said value was blank we would fall back on a default. a typical `.get(...)` was not enough for this.

Thanks for the explanation :) I'm okay with leaving as is (and sorry I didn't see your comment :P)

Reply:
I expanded on the comment and moved it further up to make it easier to see